### ISSUE TITLE
Timestamp and Mysql 5.6

### ISSUE THREAD

**Issue by atukai** [zendframework/zend-log#3]
In Logger.php

``` php
$timestamp = new DateTime();
$event = [
    'timestamp'    => $timestamp,
    'priority'     => (int) $priority,
    'priorityName' => $this->priorities[$priority],
    'message'      => (string) $message,
    'extra'        => $extra,
];
```

Uncaught exception 'PDOException' with message 'SQLSTATE[22007]: Invalid datetime format: 1292 Incorrect datetime value: '2015-06-19T18:44:10+03:00' for column.

Maybe we need?

``` php
$timestamp->format(Y-m-d H:i:s)
```

**Comment by weierophinney** [comment]
No — because that already happens.

The logger instances creates the event with a `DateTime` instance, but then individual writers compose a `Zend\Log\Formatter\FormatterInterface` for formatting the event and/or elements of the event into something they can write. In the case of the `Db` writer, it uses `Zend\Log\Formatter\Db`, which casts the `DateTime` instance using `$dateTime->format('c')`. You can alter the `DateTime` format provided to the formatter if necessary.

Based on your report, it looks like you've defined the 'timestamp' column of your table to use a MySQL `datetime` column type. You may need to either alter the column type, or you can alter the writer to use an alternate formatter.

If you're defining your logger via configuration, you can specify a `formatter` key to the writer configuration:

``` php
'writers' => [
    [
        'name' => 'db',
        'options' => [
            'formatter' => [
                'name' => 'db',
                'options' => [
                    'dateTimeFormat' => 'Y-m-d H:i:s',
                ],
            ],
        ],
    ],
]
```

Otherwise, you can do it manually:

``` php
foreach ($logger->getWriters() as $writer) {
    if (! $writer instanceof \Zend\Log\Writer\Db) {
        continue;
    }
    $writer->setFormatter(new \Zend\Log\Formatter\Db('Y-m-d H:i:s'));
}
```

**Comment by Martin-P** [comment]
> Based on your report, it looks like you've defined the 'timestamp' column of your table to use a MySQL `datetime` column type.

Also timestamp expects `YYYY-MM-DD HH:MM:SS`: https://dev.mysql.com/doc/refman/5.6/en/datetime.html

> TIMESTAMP has a range of '1970-01-01 00:00:01' UTC to '2038-01-19 03:14:07' UTC. 

MySQL accepts faulty dates by default, but in strict mode only valid dates are accepted and prevents non-existing dates like `0000-00-00 00:00:00`.

Also PostgreSQL expects `YYYY-MM-DD HH:MM:SS` (see [the manual](http://www.postgresql.org/docs/9.4/static/datatype-datetime.html#DATATYPE-DATETIME-INPUT)).

@weierophinney 
Why use the format `2015-06-19T18:44:10+03:00` if at least 2 major databases do not support it? Shouldn't the default format reflect the most common situation?

**Comment by codeliner** [comment]
@Martin-P With `YYYY-MM-DD HH:MM:SS` you lose the timezone information. IMO the iso 8601 format is a very good default for logging events. You just need to define the timestamp column as a varchar.

**Comment by Martin-P** [comment]
> IMO the iso 8601 format is a very good default for logging events.

If any database supports that format for `timestamp` columns, it can be good.

> You just need to define the timestamp column as a varchar.

`varchar` for dates/times? Sorry, but IMO that is a definate no go, because you can't force data integrity when using a `varchar`.

**Comment by codeliner** [comment]
> varchar for dates/times? Sorry, but IMO that is a definate no go, because you can't force data integrity when using a varchar.

Why should zend-log force data integrity on database level? The timestamp is just information and the data integrity is ensured by the code. I'm not saying that iso 8601 is the best format for all situations but for logging events it is.

**Comment by Martin-P** [comment]
> Why should zend-log force data integrity on database level?

I'm not saying `zend-log` should force data integrety on database level. I'm saying data integrety can not be forced when using `varchar`. If I put the text `3 hamburgers` in a `varchar` column the database says ok, because it is a valid `varchar` value. Such invalid values should be impossible to insert, that is why `varchar` is not an option IMO.

My point was about the default formatting of the datetime value. If 2 major databases say the format _must_ be `YYYY-MM-DD HH:II:SS` then why use a default format which does not comply with that?

**Comment by codeliner** [comment]
ok, sorry. Maybe my question was not specific enough. I try again using your example: 

> I'm saying data integrety can not be forced when using varchar

Agreed, when only looking at the database. But zend-log will always provide a valid timestamp. It is impossible that `3 hamburgers` is provided by zend-log as timestamp. This is ensured by unit tests (I hope so ...). So why not simply trust zend-log? Logging events is an append only operation. The database is only used as information storage. It doesn't need to say anything about the inserted data. I consider the timezone information as more important as 100 % data integrety on database level. At least as a default of zend-log. If you really want to make sure that no one can ever insert `3 hamburgs` well then you can change it like described by @weierophinney but be aware that you lose important information. 
But yeah, that's not my decision. Your solution would be fine for me too. Then I would change it back to iso 8601 or even `$dateTime->format('Y-m-d\TH:i:s.uO')`.

Btw. How would you store timezone information in MySql? Maybe I really do it wrong and should consider changing it.

**Comment by froschdesign** [comment]
> The database is only used as information storage.

:-1: This is not true.

You can use the database to analyze the data.
Writing is only one part of logging. Reading and analyze the other side.

**Comment by codeliner** [comment]
@froschdesign You are right. Let me rephrase: The database is normally used as information storage with regards to logging. Normally you don't perform any updates on logged events, so it is nearly impossible that another process corrupts the data.

Question is: In wich cases do you get a problem with an iso date stored as a varchar compared to a timestamp with regards to reading or analysing the logs? The iso date varchar column is sortable, you can query ranges, etc. And more important: Are these cases edge cases or should they be supported by default?

We have to POVs here:
1. The database POV: It is a no go to store dates/times in column with a type other than datetime/timestamp.
2. The application POV: It would be a good thing to have the timezone information logged because otherwise you lose important information that you may need when you want to display the log to a user located in another timezone or aggregate logs of different servers running in different timezones.

If the second one can be solved in another way: :+1: for POV 1
As long as storing the iso date as a varchar is the only way to keep the timezone: :+1: for @weierophinney suggestion to change the column type or change the default formatting -> because then you are aware that you will lose information

**Comment by Martin-P** [comment]
> so it is nearly impossible that another process corrupts the data

When you force data integrety on a database level it is not _nearly_ impossible, but _absolutely_ impossible to corrupt the data. Data integrety should always be the main goal in an application.

But perhaps we can get back to the issue and discuss why the default format does not comply with the expected format of timestamp columns of major databases?

**Comment by codeliner** [comment]
> Data integrety should always be the main goal in an application.

I agree. But in this case it is ensured by zend-log. 

Why do timestamp columns of major databases require a format without including the timezone? I don't develop applications to make the database happy I develop applications to make my users / clients happy. If the database blocks me from doing so well then I choose another database or work with a trade off.

**Comment by froschdesign** [comment]
> Why do timestamp columns of major databases require a format without including the timezone?

SQL standard

**Comment by codeliner** [comment]
@froschdesign And how can I store a ISO 8601 formated date using SQL standard?
